File: src/Commands/DetectAnomaly/CostAnalyzer.cs
using AzureCostCli.CostApi;

namespace AzureCostCli.Commands.ShowCommand;

public class CostAnalyzer
{
    private int RecentActivityDays = 7;
    private double SignificantChange = 0.5; // 50%
    private int SteadyGrowthDays = 7;
    private double ThresholdCost = 2.00d;
    
    public CostAnalyzer(DetectAnomalySettings settings)
    {
        RecentActivityDays = settings.RecentActivityDays;
        SignificantChange = settings.SignificantChange;
        SteadyGrowthDays = settings.SteadyGrowthDays;
        ThresholdCost = settings.ThresholdCost;
    }

    public List<AnomalyDetectionResult> AnalyzeCost(List<CostDailyItem> items)
    {
        var groupedItems = items
            .GroupBy(i => i.Name)
            .Select(g => g.OrderBy(i => i.Date).ToList())
            .ToList();

        var results = new Dictionary<(string, AnomalyType), AnomalyDetectionResult>();
        foreach (var group in groupedItems)
        {
            AnomalyDetectionResult result;

            // Check for new costs
            if (group.First().Cost == 0 && group.Skip(1).Any(i => i.Cost != 0 && i.Cost > ThresholdCost))
            {
                var startCostItem = group.First(i => i.Cost != 0);
                result = new AnomalyDetectionResult
                {
                    Name = startCostItem.Name,
                    DetectionDate = startCostItem.Date,
                    Message = "New cost detected at " + startCostItem.Date,
                    CostDifference = startCostItem.Cost,
                    AnomalyType = AnomalyType.NewCost,
                    Data = group
                };
                results[(startCostItem.Name, AnomalyType.NewCost)] = result;
            }

            // Check for removed costs
            if (group.Last().Cost == 0)
            {
                for (int i = group.Count - 2; i >= 0; i--)
                {
                    if (group[i].Cost != 0)
                    {
                        var endCostItem = group[i];
                        var detectionDate = endCostItem.Date.AddDays(1);
                        result = new AnomalyDetectionResult
                        {
                            Name = endCostItem.Name,
                            DetectionDate = detectionDate, // Assuming the cost was removed the next day
                            Message = $"Cost is removed at {detectionDate}",
                            CostDifference = -endCostItem.Cost,
                            AnomalyType = AnomalyType.RemovedCost,
                            Data = group
                        };
                        results[(endCostItem.Name, AnomalyType.RemovedCost)] = result;
                        break;
                    }
                }
            }

            // Filter out inactive resources
            if (!IsResourceActive(group)) continue;

            // Check for significant cost changes
            for (int i = 1; i < group.Count; i++)
            {
                var today = group[i];
                var yesterday = group[i - 1];
                var diff = today.Cost - yesterday.Cost;

                if (Math.Abs(diff) / yesterday.Cost > SignificantChange && diff > ThresholdCost)
                {
                    result = new AnomalyDetectionResult
                    {
                        Name = today.Name,
                        DetectionDate = today.Date,
                        Message = $"Significant cost change detected at {today.Date} when it went from {yesterday.Cost:N2} to {today.Cost:N2} (difference of {diff:N2})",
                        CostDifference = diff,
                        AnomalyType = AnomalyType.SignificantChange,
                        Data = group
                    };
                    results[(today.Name, AnomalyType.SignificantChange)] = result;
                }
            }

            // Check for steady cost increase over a week
            if (IsSteadyCostIncrease(group))
            {
                var today = group.Last();
                result = new AnomalyDetectionResult
                {
                    Name = today.Name,
                    DetectionDate = today.Date,
                    Message = "Steady cost increase over a week detected",
                    CostDifference = today.Cost - group[^SteadyGrowthDays].Cost,
                    AnomalyType = AnomalyType.SteadyGrowth,
                    Data = group
                };
                results[(today.Name, AnomalyType.SteadyGrowth)] = result;
            }
        }

        return results.Values.ToList();
    }

    private bool IsResourceActive(IEnumerable<CostDailyItem> items)
    {
        return items.Any(i => i.Date >= DateOnly.FromDateTime(DateTime.Now.AddDays(-RecentActivityDays)));
    }

    private bool IsSteadyCostIncrease(IReadOnlyCollection<CostDailyItem> items)
    {
        if (items.Count < SteadyGrowthDays) return false;

        var lastWeekItems = items.TakeLast(SteadyGrowthDays).ToList();
        for (int i = 1; i < lastWeekItems.Count; i++)
        {
            if (lastWeekItems[i].Cost <= lastWeekItems[i - 1].Cost) return false;
        }

        return true;
    }
}

File: python/azure_cost_cli/commands/cost_analyzer.py
"""Anomaly detection helper – mirrors CostAnalyzer.cs."""
from __future__ import annotations

from datetime import date, datetime, timedelta, timezone
from typing import Any

from azure_cost_cli.models import AnomalyDetectionResult, AnomalyType, CostDailyItem


def analyze_cost(
    items: list[CostDailyItem],
    recent_activity_days: int = 7,
    significant_change: float = 0.75,
    steady_growth_days: int = 7,
    threshold_cost: float = 2.0,
    exclude_removed_costs: bool = False,
) -> list[AnomalyDetectionResult]:
    from itertools import groupby as _groupby

    today = datetime.now(timezone.utc).date()
    cutoff = today - timedelta(days=recent_activity_days)

    grouped: dict[str, list[CostDailyItem]] = {}
    for item in sorted(items, key=lambda x: x.date):
        grouped.setdefault(item.name, []).append(item)

    results: dict[tuple, AnomalyDetectionResult] = {}

    for name, group in grouped.items():
        group = sorted(group, key=lambda x: x.date)

        # New cost
        if group[0].cost == 0 and any(i.cost != 0 and i.cost > threshold_cost for i in group[1:]):
            start = next(i for i in group if i.cost != 0)
            results[(start.name, AnomalyType.NEW_COST)] = AnomalyDetectionResult(
                name=start.name,
                detection_date=start.date,
                message=f"New cost detected at {start.date}",
                cost_difference=start.cost,
                anomaly_type=AnomalyType.NEW_COST,
                data=group,
            )

        # Removed cost
        if not exclude_removed_costs and group[-1].cost == 0:
            for i in range(len(group) - 2, -1, -1):
                if group[i].cost != 0:
                    end = group[i]
                    det_date = end.date + timedelta(days=1)
                    results[(end.name, AnomalyType.REMOVED_COST)] = AnomalyDetectionResult(
                        name=end.name,
                        detection_date=det_date,
                        message=f"Cost is removed at {det_date}",
                        cost_difference=-end.cost,
                        anomaly_type=AnomalyType.REMOVED_COST,
                        data=group,
                    )
                    break

        # Active resource check
        if not any(i.date >= cutoff for i in group):
            continue

        # Significant change
        for i in range(1, len(group)):
            today_item = group[i]
            yesterday_item = group[i - 1]
            if yesterday_item.cost == 0:
                continue
            diff = today_item.cost - yesterday_item.cost
            if abs(diff) / yesterday_item.cost > significant_change and diff > threshold_cost:
                results[(today_item.name, AnomalyType.SIGNIFICANT_CHANGE)] = AnomalyDetectionResult(
                    name=today_item.name,
                    detection_date=today_item.date,
                    message=(
                        f"Significant cost change detected at {today_item.date} when it went from "
                        f"{yesterday_item.cost:.2f} to {today_item.cost:.2f} (difference of {diff:.2f})"
                    ),
                    cost_difference=diff,
                    anomaly_type=AnomalyType.SIGNIFICANT_CHANGE,
                    data=group,
                )

        # Steady growth
        if len(group) >= steady_growth_days:
            last_n = group[-steady_growth_days:]
            if all(last_n[j].cost > last_n[j - 1].cost for j in range(1, len(last_n))):
                last = group[-1]
                results[(last.name, AnomalyType.STEADY_GROWTH)] = AnomalyDetectionResult(
                    name=last.name,
                    detection_date=last.date,
                    message="Steady cost increase over a week detected",
                    cost_difference=last.cost - group[-steady_growth_days].cost,
                    anomaly_type=AnomalyType.STEADY_GROWTH,
                    data=group,
                )

    return list(results.values())
